elfpy/markets/hyperdrive/hyperdrive_market.py

```diff
@@ -443,3 +517,80 @@ def remove_liquidity(
         self.market_state.apply_delta(market_deltas)
         agent_wallet.update(agent_deltas)
         return market_deltas, agent_deltas
+
+    def checkpoint(self, checkpoint_time: float):
+        """allows anyone to mint a new checkpoint."""
+        # if the checkpoint has already been set, return early.
+
+        if self.market_state.checkpoints[checkpoint_time].share_price != 0:
+            return
+
+        # if the checkpoint time isn't divisible by the checkpoint duration
+        # or is in the future, it's an invalid checkpoint and we should
+        # revert.
+        latest_checkpoint = self.get_latest_checkpoint_time()
+        # TODO: this modulo check should be just > 0, but floats are weird
+        if checkpoint_time % self.market_state.checkpoint_duration > 1e17 or latest_checkpoint < checkpoint_time:
+            raise errors.InvalidCheckpointTime()
+
+        # if the checkpoint time is the latest checkpoint, we use the current
+        # share price. otherwise, we use a linear search to find the closest
+        # share price and use that to perform the checkpoint.
+        if checkpoint_time == latest_checkpoint:
+            self.apply_checkpoint(latest_checkpoint, self.market_state.share_price)
+        else:
+            _time = checkpoint_time
+            while True:
+                closest_share_price = self.market_state.checkpoints[_time].share_price
+                if _time == latest_checkpoint:
+                    closest_share_price = self.market_state.share_price
+                if closest_share_price != 0:
+                    self.apply_checkpoint(checkpoint_time, closest_share_price)
+                    break
+                _time += self.market_state.checkpoint_duration
+
+    def get_latest_checkpoint_time(self):
```

Comment:
I would prefer for this to be a property, so that we can invoke it with `market.latest_checkpoint_time`.